lib/pulsar/producer.ex

```diff
@@ -254,18 +311,48 @@ defmodule Pulsar.Producer do
     end
   end
 
+  @impl true
+  def handle_info({:broker_message, %Binary.CommandProducerSuccess{} = command}, state) do
+    if state.registration_request_id == command.request_id do
+      # Save topic_epoch if we received one (for ExclusiveWithFencing producers)
+      if command.topic_epoch do
+        ProducerEpochStore.put(state.topic, state.producer_name, state.access_mode, command.topic_epoch)
+      end
+
+      {:noreply, %{state | ready: command.producer_ready, topic_epoch: command.topic_epoch}}
+    else
+      {:noreply, state}
+    end
+  end
+
+  @impl true
+  def handle_info({:broker_message, %Binary.CommandCloseProducer{}}, state) do
+    {:stop, :broker_close_requested, state}
+  end
+
+  @impl true
+  def handle_info(:retry_registration, state) do
+    # Retry registration by triggering continue
+    {:noreply, state, {:continue, :register_producer}}
+  end
+
   @impl true
   def terminate(_reason, nil) do
     :ok
   end
 
   @impl true
-  def terminate(_reason, state) do
+  def terminate(reason, state) do
     Logger.debug("Terminating producer: #{inspect(state.producer_name)}")
 
+    # Clean up epoch entry if producer is being permanently stopped (fenced or normal shutdown)
+    if reason in [:shutdown, {:shutdown, :producer_fenced}, :normal] do
+      ProducerEpochStore.delete(state.topic, state.producer_name, state.access_mode)
```

Comment:
I am worried that having this in the `terminate/2` callback function may lead to data not being cleaned up, since Erlang/Elixir do not guarantee that `terminate` will be executed in full.